### PR TITLE
Fix Storage lifetime in table functions

### PR DESCRIPTION
**Pull request by tavplubix**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Bug Fix

Changelog entry (up to few sentences, required except for Non-significant/Documentation categories):
Fix possible crashes in table functions (`file`, `mysql`, `remote`) caused by usage of reference to removed `IStorage` object. Fix incorrect parsing of columns specified at insertion into table function.

Detailed description (optional):
`InterpreterSelectQuery` has own `StoragePtr` and `Context`. Some `Streams` in query pipeline may have references to those storage or context, but interpreter with storage (if it's temporary storage created by table function) and context is destructed before pipeline execution. 
Added shared pointers to storage and interpreter context in query pipeline

### CODE REVIEW

File: dbms/src/Processors/QueryPipeline.cpp [hunk 1]
```diff
@@ -486,6 +486,10 @@ void QueryPipeline::unitePipelines(
 
         processors.insert(processors.end(), pipeline.processors.begin(), pipeline.processors.end());
         streams.insert(streams.end(), pipeline.streams.begin(), pipeline.streams.end());
+
+        interpreter_context.insert(interpreter_context.end(), pipeline.interpreter_context.begin(), pipeline.interpreter_context.end());
+        storage_holder.insert(storage_holder.end(), pipeline.storage_holder.begin(), pipeline.storage_holder.end());
+        /// Do we need also copy table_locks?
```
Comment (KochetovNicolai) [inline dbms/src/Processors/QueryPipeline.cpp]: Yes. It is also better to move locks from other pipelines.

File: dbms/src/Interpreters/InterpreterInsertQuery.cpp [hunk 2]
```diff
@@ -98,7 +98,10 @@ BlockIO InterpreterInsertQuery::execute()
 {
     const auto & query = query_ptr->as<ASTInsertQuery &>();
     checkAccess(query);
+
+    BlockIO res;
     StoragePtr table = getTable(query);
+    res.pipeline.addStorageHolder(table);
```
Comment (KochetovNicolai) [inline dbms/src/Interpreters/InterpreterInsertQuery.cpp]: Funny thing that we use pipeline only like lock holder here. But it's ok.

**Review by KochetovNicolai — APPROVED** [review]
There is also fix in parser for insert into table functions. Better mention it in changelog.

**Comment by nikitamikhaylov** [comment]
Conflicts with 19.16

**Pull request merged**